File: NidaAktan/soru2.js
/**
* // Definition for a Node.
* function Node(val, left, right, next) {
* this.val = val === undefined ? null : val;
* this.left = left === undefined ? null : left;
* this.right = right === undefined ? null : right;
* this.next = next === undefined ? null : next;
* };
*/
/**
* @param {Node} root
* @return {Node}
*/
  function Node(val, left, right, next) {
 this.val = val === undefined ? null : val;
 this.left = left === undefined ? null : left;
 this.right = right === undefined ? null : right;
 this.next = next === undefined ? null : next;
};
   var connect = function (root) {
        if (!root) {
            return null;
        }
        if (root.next && root.right) {
            root.next.left ? root.right.next = root.next.left : root.right.next = root.next.right;
        }
        if (root.left && root.right) {
            root.left.next = root.right;
        }
        connect(root.left);
        connect(root.right);
        return root;
    };
    var root = new Node(1);
    root.left = new Node(2);
    root.left.right = new Node(5);
    root.right = new Node(3);
    root.right.left = null;
    root.left.left = new Node(4);
    root.right.right = new Node(7);

File: AhmetBaranBogurcu/Soru 2.java
package com.bites;

public class Main {

    public static void main(String[] args) {
        RightToBinaryTreeImpl rightToBinaryTree = new RightToBinaryTreeImpl();
        rightToBinaryTree.connect(new Node());
    }
}

class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {
    }

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
}

class RightToBinaryTreeImpl implements RightToBinaryTree {

    @Override
    public Node connect(Node root) {
        if (root == null)
            return null;

        Node node = root.next;
        Node nextNode = null;

        while (node != null) {
            if (node.left != null) {
                nextNode = node.left;
                break;
            }
            if (node.right != null) {
                nextNode = node.right;
                break;
            }

            node = node.next;
        }

        if (root.right != null)
            root.right.next = nextNode;
        if (root.left != null)
            root.left.next = root.right == null ? nextNode : root.right;

        connect(root.right);
        connect(root.left);

        return root;
    }
}

interface RightToBinaryTree {
    public Node connect(Node root);
}
